Topic: secret management (#16)
Nietecht/mille-feuille-opinions: topics/secret-management/index.md

Using Azure Key Vault like this covers most cases. You still have to account for different environments and authentication scenario's, but this is the gist of it.  

### Adding Azure App Configuration to the mix

Integrating Azure Key Vault into Azure App Configuration is easy and replaces the Key Vault integration mentioned above.

``` csharp
var cred = new DefaultAzureCredential();

return config.AddAzureAppConfiguration(options =>
{
    options
        .Connect(appConfigEndpoint, cred)
        .ConfigureKeyVault(kv =>
        {
            kv.SetCredential(cred);
        })
        .Select(KeyFilter.Any, LabelFilter.Null)
        .Select(KeyFilter.Any, environmentName);
});
```

Requires `Microsoft.Extensions.Configuration.AzureAppConfiguration` and `Azure.Identity` {.code-annotation}

> `DefaultAzureCredential` works a bit different than `AzureServiceTokenProvider` however. The most important difference I encountered is that when using an Azure Service Authentication account in Visual Studio that is a guest account to the Azure directory your resources are in, you need to explicitly configure the tenant id as well as use at least version `1.2.0-preview.2` of `Azure.Identity`.  

Reviewer: 1. Is this only true in visual studio or also in when you have this service deployed?
2. Is this only true vs the Azure App Configuration resource, or any resource you want to authenticate against (like Key Vault)

> Just as with `DefaultAzureCredential`, you also need to configure the client id when working with user-assigned managed identity.

Reviewer: ```suggestion
Just as with `AzureServiceTokenProvider`, you also need to configure the client id when working with user-assigned managed identity.
```

``` csharp
var cred = new DefaultAzureCredential(new DefaultAzureCredentialOptions
{
    VisualStudioTenantId = tenantId,
    VisualStudioCodeTenantId = tenantId,
    SharedTokenCacheTenantId = tenantId,
    InteractiveBrowserTenantId = tenantId,
    ManagedIdentityClientId = userManagedIdentityClientId
});
```

Configuring `VisualStudioTenantId` and `VisualStudioCodeTenantId` requires `Azure.Identity` version `1.2.0-preview.2` or higher. {.code-annotation}

You might wonder, if this is during the setup of our configuration, how do we get all those variables that need to be be, well, configured? Sounds a bit like a chicken and egg problem.  

> A simple trick is the solution: you can call `.Build()` on the `IConfigurationBuilder` object to get an `IConfiguration` object you can use while still configuring the original `IConfigurationBuilder`. Of course, you need to make sure there's already a source configured, usually simply the configuration file or environment variables.

Reviewer: ```suggestion
A simple trick is the solution: you can call `.Build()` on the `IConfigurationBuilder` object to get an `IConfiguration` object you can use while still configuring the original `IConfigurationBuilder`. Of course, you need to make sure there's already a source configured, usually the normal `appsettings.json` configuration file or the environment variables.
```

### Azure Functions

> This approach works for Azure Functions as well. However, as of writing, for Azure Functions V3, `Microsoft.Extensions.DependencyInjection.Abstractions` (and related packages) require version `3.1.0` instead of the latest version. This is a bug that should get fixed, so check if things don't just work with the newest version anyway.

Reviewer: You have this bug somewhere? It should be linked here!

## Practical examples

> For ease of use, I've created a couple of utility methods that cover my use cases.

Reviewer: ```suggestion
For ease of use, I've created a couple of utility methods that cover these use cases.
```
or `our` would also be better and more inclusive

> The first method, `AddAzureAppConfig`, configures Azure App Configuration and Azure Key Vault. It configures the tenant id for all authentication strategies and reads the App Configuration values in the standard way: first read all values without a label and then all those with the environment name as label, overriding those without label as needed.

Reviewer: ```suggestion
The first method, `AddAzureAppConfig`, configures Azure App Configuration and Azure Key Vault. It configures the tenant id for all authentication strategies and reads the App Configuration values in the standard way: first read all values without a label and then all those with the environment name as label, overriding those without label as needed.
```
```suggestion
The first method, `AddAzureAppConfig`, configures Azure App Configuration and Azure Key Vault. It configures the tenant id for all authentication strategies and reads the App Configuration values in the standard way. First read all values without a label. Then read all those with the environment name as label, overriding those without label as needed.
```

    // Override with dev appsettings allows easy configuring of non-sensitive settings without having to put them in user secrets, so they can be shared
    if (env.IsDevelopment()) {
        config = config
            .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true, reloadOnChange: true);
    }

    return config
        .AddUserSecrets<TSecretAssemblyClass>()
        .AddEnvironmentVariables();
}
```

### Usage in a web service

Using this in a web service is simple enough, just call it when building your host object.

``` csharp
Host.CreateDefaultBuilder(args)
  .ConfigureAppConfiguration((context, config) => config.AddDefaultAppConfig<Startup>(context.HostingEnvironment))
  .ConfigureWebHostDefaults(webHostBuilder => webHostBuilder.UseStartup<Startup>())
  .Build();
```

Note that using the default host builder, the default `appsettings.json` files are already configured as a source, which is why `AddDefaultAppConfig` can read the values it needs to set up the rest of the configuration. If you don't use the default builder, you might need to add those sources yourself first, or devise another way to get the required values.

> `appsettings.json` might then look like this:

Reviewer: 1. This you might have to override on deployment, when using other Tenants or Azure App Configurations for production
2. Don't forget to add `UserManagedIdentityClientId` if you need it
